Make class properties observable

cherifGsoul opened the pull request:
This makes class properties observable:
```js
class MyElement extends StacheElement {
		greetings = 'Hello';
                static get view() { return {{ greetings }}}
		static get props() {
		}
}
customElements.define('my-element', MyElement)
```
In case of:

```js
class MyElement extends StacheElement {
		greetings = 'Hello';
                static get view() { return {{ greetings }}}
		static get props() {
                       return {
				greetings: {type: String, default: 'Bonjour'}
			};
		}
}
customElements.define('my-element', MyElement)
```
It set the value of `props.greetings`  by the classe property value.

For https://github.com/canjs/can-observable-mixin/issues/79

cherifGsoul committed:
Make classe properties observable

cherifGsoul committed:
Ignore fs-events in detect-cyclic-packages

cherifGsoul committed:
remove babel plugin-proposal-class-properties

cherifGsoul committed:
Skip test on non-supported browsers

phillipskevin commented on src/mixin-props-test.js:
```diff
@@ -20,3 +20,61 @@ QUnit.test("basics", function(assert) {
 	assert.equal(el.age, 33, "updated age");
 });
+
+try {
```
I think all of the tests that require class fields should go in their own file. Then the main test file can do this feature detection and create a variable like `supportsClassFields`. The wrap the `require` statement for the tests like

```
if(supportsClassFields) {
  require(...);
}
```

phillipskevin commented on src/mixin-props-test.js:
```diff
@@ -20,3 +20,61 @@ QUnit.test("basics", function(assert) {
+		QUnit.test('Class fields should be observable', (assert) => {
+			class DefineElement extends mixinDefine(Object) {
+				greetings = 'Hello'; // jshint ignore:line
+				static get props() {
```
Are `props` required?

phillipskevin commented on src/mixin-props-test.js:
```diff
@@ -20,3 +20,61 @@ QUnit.test("basics", function(assert) {
+			const el = new DefineElement();
+
+			if (el.hasOwnProperty('greetings')) {
+				el.initialize();
```
You should probably check that the default value is correct.

phillipskevin commented on src/mixin-props-test.js:
```diff
@@ -20,3 +20,61 @@ QUnit.test("basics", function(assert) {
+				el.initialize();
+			
+				el.on('greetings', () => {
+					assert.ok('The class field is observable');
```
Can you also check that the value and old value are correct?

phillipskevin commented on src/mixin-props-test.js:
```diff
@@ -20,3 +20,61 @@ QUnit.test("basics", function(assert) {
+		}
+	
+		const el = new DefineElement();
+			el.initialize();
```
You also need to check the value of `greetings` to make sure the precedence is correct.

phillipskevin commented on src/mixin-props-test.js:
```diff
@@ -20,3 +20,61 @@ QUnit.test("basics", function(assert) {
+				assert.ok('The class field is observable');
+			});
+	
+			el.greetings = 'Hola';
```
After this you should also check that setting the value to a non-string causes an error.

phillipskevin commented on src/mixin-props-test.js:
```diff
@@ -20,3 +20,61 @@ QUnit.test("basics", function(assert) {
+					assert.ok('The class field is observable');
+				});
+			
+				el.greetings = 'Hola';
```
Can you also check that setting the value to a non-string works since expando properties like this are not typed.

phillipskevin commented on src/mixin-props.js:
```diff
@@ -56,8 +56,26 @@ module.exports = function mixinDefine(Base = HTMLElement) {
+				if (this.hasOwnProperty(prop)) {
+					if (staticProps && staticProps.includes(prop)) {
+						const val = this[prop];
+						delete this[prop];
```
Why do you need to delete the prop?

cherifGsoul replied:
Because `this[prop]` will take the class field property with the same name.
`this.constructor.props[prop]` is not settable.

phillipskevin reviewed (changes requested):
Good start @cherifGsoul!

cherifGsoul committed:
Put the tests in their own file

cherifGsoul committed:
Change to import

cherifGsoul committed:
Organie the tests

cherifGsoul committed:
add another test

cherifGsoul committed:
Remove useless test

cherifGsoul committed:
Add documentation for class fields

phillipskevin commented on docs/can-stache-element.md:
```diff
@@ -206,6 +206,31 @@ document.body.querySelector("button#remove").addEventListener("click", () => {
 @highlight 14-23,only
 
+
+#### Observable class fields
```
Can you move this under the `Defining an element's properties` heading?

cherifGsoul replied:
Done!

phillipskevin commented on docs/can-stache-element.md:
```diff
@@ -206,6 +206,31 @@ document.body.querySelector("button#remove").addEventListener("click", () => {
+
+#### Observable class fields
+
+`StachElement` class fields are observables like [can-stache-element/static.props static props],
```
Can you link `class fields` to https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_fields?

cherifGsoul replied:
Done!

phillipskevin commented on docs/can-stache-element.md:
````diff
@@ -206,6 +206,31 @@ document.body.querySelector("button#remove").addEventListener("click", () => {
+</script>
+```
+@codepen
+@highlight 11,only
````
The highlighting is off here.

cherifGsoul replied:
Fixed!

phillipskevin commented on docs/can-stache-element.md:
```diff
@@ -206,6 +206,31 @@ document.body.querySelector("button#remove").addEventListener("click", () => {
+    <button on:click="this.increment()">+1</button>
+  `;
+  
+  count: 6
```
This should be `count = 6;`.

cherifGsoul replied:
Done!

phillipskevin reviewed (changes requested):
Added some comments on the docs.

Will review the code separately.

phillipskevin commented on docs/can-stache-element.md:
```diff
@@ -206,6 +206,31 @@ document.body.querySelector("button#remove").addEventListener("click", () => {
+  count: 6
+}
+customElements.define("count-er", Counter);
+</script>
```
Can you also add a note in that you must either add the element to the page or call `initialize` in order for the `prop` value to be observable?

cherifGsoul replied:
Done!

cherifGsoul committed:
Update class fields docs

cherifGsoul commented:
@phillipskevin I updated the docs!

cherifGsoul committed:
Fix link

phillipskevin commented on docs/can-stache-element.md:
```diff
@@ -135,6 +135,33 @@ customElements.define("count-er", Counter);
+@codepen
+@highlight 11,only
+
+In order to make `count` property observable, either `count-er` element must be added to the page
```
Can you put this in a blockquote? `>` and add `Note: ` to the beginning?

phillipskevin commented on docs/can-stache-element.md:
```diff
@@ -135,6 +135,33 @@ customElements.define("count-er", Counter);
 
+#### Observable class fields
+
+`StachElement` [class fields](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes/Class_fields) like [can-stache-element/static.props static props],
```
I think this is missing some words. 

> class fields are observable like...

phillipskevin replied:
Also, I would just leave off the `like static props` piece. Just

> StacheElement class fields are also observable.

...along with the example should be fine for now.

phillipskevin reviewed (changes requested):
Just a couple small comments.

phillipskevin commented on test/test.js:
```diff
@@ -7,3 +18,9 @@ import "../src/import-export-steal-test";
+if (supportsClassFields) {
+	//It doesn't work with require
+	//Even when change the above imports to require
+	steal.import('~/src/mixin-props-class-fields-test');
```
I just tried this out in the canjs/canjs suite and this file can't be loaded since the `~` is the `canjs` folder. We'll need to figure out how to do this.

phillipskevin reviewed (changes requested):
Found one more issue.

cherifGsoul committed:
Update how to require/import test

cherifGsoul committed:
Fix dynamic importing

cherifGsoul committed:
Use steal-conditional for safari

phillipskevin reviewed (changes requested):
It looks like you added steal-conditional and created the supports-class-fields module, but it is not being used to import the class field test file.

Also, there are still a couple issues with the docs that need to be cleaned up.

cherifGsoul committed:
Commit test changes

cherifGsoul committed:
Update docs

cherifGsoul commented:
> It looks like you added steal-conditional and created the supports-class-fields module, but it is not being used to import the class field test file.
> 
> Also, there are still a couple issues with the docs that need to be cleaned up.

@phillipskevin I updated the docs
The test file was not in the commit, I added it.

phillipskevin commented on docs/can-stache-element.md:
```diff
@@ -135,6 +135,33 @@ customElements.define("count-er", Counter);
 
+#### Observable class fields
+
+`StachElement` class fields are also observable.,
```
can you add the link to class fields docs on MDN?

cherifGsoul replied:
@phillipskevin Done!

phillipskevin commented on test/test.js:
```diff
@@ -1,3 +1,14 @@
+let supportsClassFields;
```
This can be removed now, right?

cherifGsoul replied:
@phillipskevin Done!

phillipskevin reviewed (changes requested):
A couple more small things.

cherifGsoul committed:
Minor changes

phillipskevin reviewed (approved):
🎉
